### DISCUSSION TITLE
Incorporate BTT Knomi macros

### DISCUSSION THREAD

**Discussion by appleimperio** [jschuh/klipper-macros#137] [Q&A] [answered]
How can I use the BTT Knomi macros for homing and Leveling so I can see the status in the Knomi UI https://bigtreetech.github.io/docs/KNOMI.html# Thank You

````
[gcode_macro BED_MESH_CALIBRATE]
rename_existing: BED_MESH_CALIBRATE_BASE
variable_probing:False

gcode:
     SET_GCODE _VARIABLE MACRO=BED_MESH_CALIBRATE VARIABLE=probing VALUE=True
     BED_MESH_CALIBRATE_BASE
     SET_GCODE_VARIABLE MACRO=BED_MESH_CALIBRATE VARIABLE=probing VALUE=False

[gcode_macro G28]
rename_existing: G0028
variable_homing:False

 gcode:
     SET_GCODE_VARIABLE MACRO=628 VARIABLE-homing VALUE=True
     G0028
     SET_GCODE_VARIABLE MACRO=628 VARIABLE=homing VALUE=False
````

**Comment by appleimperio** [comment, marked as answer]
There is a better firmware from https://github.com/Surion79/KNOMI that provides a better way to handle the homing and probing status that don't mess with the g28 or bed_mesh_calibrate

**Comment by jschuh** [comment]
Those should just work, although you may need to put them in a separate file that gets included last (to ensure Klipper wraps the macros in the correct order). Also, there does seem to be a typo where `VARIABLE-homing` should be `VARIABLE=homing`.

If the KNOMI provides a way to directly set the status, then it would probably be better use [printer status events](https://github.com/jschuh/klipper-macros#print-status-events). But I didn't dig deep enough to see if that would work.

**Reply by appleimperio** [reply]
Yes with the new firmware they provide a knomi_status and as you suggest I moved them to the Printer Status Event. I'm see you add a status led example I'm trying to figure it out to make it work with WLED. This Macros keep getting better. Thank You!

**Reply by jschuh** [reply]
Oh, I see. There's a KNOMI_STATUS_Macro branch. Unfortunately, that doesn't add any functionality. It just [moves the two variables to a different macro](https://github.com/bigtreetech/KNOMI/compare/d2236f471eb0261e48b65e1df943b93bf09d670b...Surion79:KNOMI:KNOMI_STATUS_Macro). So, it doesn't seem worth bothering with.

I did get curious and peek at the firmware. AFAICT it just connects to Moonraker over wifi, then runs a loop that continuously polls state from [the REST API](https://github.com/bigtreetech/KNOMI/blob/d2236f471eb0261e48b65e1df943b93bf09d670b/Firmware/src/main.cpp#L981-L988) and updates the display with the results.

So, it just doesn't do much that you can control. If I were making something like this, I'd subscribe to the relevant Moonraker events (including a remote method registration). That would avoid hammering the client with the polling traffic, and provide for customization via macros, etc.

I also doubt I'd use wifi given that it's physically attached to the printer, but I assume they wanted avoided having to change anything on the Klipper host.
